[GarbageCollector] update user guide

caesarxuchao opened the pull request:
cc @kubernetes/docs

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubernetes/kubernetes.github.io/1152)
<!-- Reviewable:end -->

caesarxuchao commented:
@mbohlool friendly ping :)

caesarxuchao commented:
Thanks @mbohlool. @devin-donnelly do I need more label to get it merged?

mbohlool commented:
I guess This PR may need a DOC LGTM (by a tech writer). If that is the case, I would suggest we send these type of changes initiated by developers directly to tech writers as they make sense to an engineer (the PR owner) and validating the truth of the statement is not part of another developer's review. There is not much value in having another developer review these changes, having a tech writer as reviewer should be enough.

caesarxuchao commented:
#1159 will fix the travis failure

devin-donnelly commented:
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[_data/guides.yml, line 61 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRA9m5ZmuDSyLReh3kA:-KRA9m5ZmuDSyLReh3kB:bs3h6oy) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/_data/guides.yml#L61)):*
> ```YAML
>   - title: Using kubectl to Manage Resources
>     path: /docs/user-guide/working-with-resources/
>   - title: Garbage collection (Beta)
> ```

Capitalize the "C" in "collection

---

*[docs/user-guide/garbage-collector.md, line 11 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRA9_FTqsZWQx8iwY_0:-KRA9_FUMG7mPVRE0AJG:bobm3g8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L11)):*
> ```Markdown
> {:toc}
> 
> ## Garbage Collector is a beta feature and is enabled by default
> ```

Is this the main header? If so, change to just "Garbage Collection".

Afterward add a paragraph: "Note: the Garbage Collector is a beta feature and enabled by default in Kubernetes version 1.4."

---

*[docs/user-guide/garbage-collector.md, line 13 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRA9qK9NZkxXcIG6x6Z:-KRA9qK9NZkxXcIG6x6_:b-a1e8p8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L13)):*
> ```Markdown
> ## Garbage Collector is a beta feature and is enabled by default
> 
> ### What is garbage collector for
> ```

Change heading to "What Does Garbage Collection Do?"

---

*[docs/user-guide/garbage-collector.md, line 15 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAA22uZA9cIwoVkF6e:-KRAA22uZA9cIwoVkF6f:bglgjbx) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L15)):*
> ```Markdown
> ### What is garbage collector for
> 
> When a replica set is deleted, in many cases it is desirable that server will automatically garbage collects all the pods created by the replica set. The garbage collector (GC) implements this. Generally, when an owner is deleted, GC deletes its dependents.
> ```

Passive voice.

"When you delete a replica set, it is often desirable for the server [editorial note: should this be "Master"?] to automatically garbage collect all the pods that the replica set has created. When you delete an owner object, the Garbage Collector (GC) deletes that owner's dependent objects.

---

*[docs/user-guide/garbage-collector.md, line 17 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAAoMmdNrH1TusZilR:-KRAAoMmdNrH1TusZilS:b-n5vupt) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L17)):*
> ```Markdown
> When a replica set is deleted, in many cases it is desirable that server will automatically garbage collects all the pods created by the replica set. The garbage collector (GC) implements this. Generally, when an owner is deleted, GC deletes its dependents.
> 
> ### How to establish owner-dependent relation
> ```


"How to establish an owner-dependent relationship between objects"

---

*[docs/user-guide/garbage-collector.md, line 19 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAAuZlRod0r7YTLZPd:-KRAAuZlRod0r7YTLZPe:b-79toy5) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L19)):*
> ```Markdown
> ### How to establish owner-dependent relation
> 
> In release 1.3, a `metadata.ownerReferences` field is introduced to every Kubernetes API objects. When an object has an `ownerReference` pointing to another object, the dependent-owner relation is established.
> ```

More passive voice.

"Kubernetes added 1.3 added a `metadata.ownerReferences` field to every Kubernetes API object. If an API object is a dependent of another object, `ownerReference` should point to the owning API object.

---

*[docs/user-guide/garbage-collector.md, line 21 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRABJjY2kGhHlHw251q:-KRABJjY2kGhHlHw251r:b-c77b1e) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L21)):*
> ```Markdown
> In release 1.3, a `metadata.ownerReferences` field is introduced to every Kubernetes API objects. When an object has an `ownerReference` pointing to another object, the dependent-owner relation is established.
> 
> In release 1.4, Kubernetes control plane will automatically set a pod to have an ownerReference pointing to the replication controller or the replica set that creates it.
> ```

"When you create a Replication Controller or ReplicaSet in Kubernetes 1.4, the Kubernetes control plane automatically sets the `ownerReference` field for each pod that set creates.

---

*[docs/user-guide/garbage-collector.md, line 23 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRACsRKajoZ1X39GpZG:-KRACsRKajoZ1X39GpZH:bhcipc8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L23)):*
> ```Markdown
> In release 1.4, Kubernetes control plane will automatically set a pod to have an ownerReference pointing to the replication controller or the replica set that creates it.
> 
> To set up owner-dependent relation among other objects, you can manually update the `ownerReferences` field of the dependent objects.
> ```

You can set up owner-dependent relationships among other objects in the Kubernetes API by manually setting the `ownerReference` field on dependent objects.

---

*[docs/user-guide/garbage-collector.md, line 25 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAD494QUqvEb_umrIh:-KRAD494QUqvEb_umrIi:b-lntkst) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L25)):*
> ```Markdown
> To set up owner-dependent relation among other objects, you can manually update the `ownerReferences` field of the dependent objects.
> 
> ### How to request the garbage collector to delete or not delete dependents
> ```

Change to "Controlling how Garbage Collector deletes dependents"

---

*[docs/user-guide/garbage-collector.md, line 27 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRADAPFlYywDXx7Rbf6:-KRADAPFlYywDXx7Rbf7:b-owyrpf) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L27)):*
> ```Markdown
> ### How to request the garbage collector to delete or not delete dependents
> 
> When deleting an object, you can request the GC to ***asynchronously*** delete its dependents by ***explicitly*** specifying `deleteOptions.orphanDependents=false` in the deletion request sent to the API server. A 200 OK response from the API server indicates the owner is deleted.
> ```

"in the deletion request that you send to the API server."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152)*
<!-- Sent from Reviewable.io -->

devin-donnelly commented:
Some writerly comments. Once those are cleared up, we can merge.

caesarxuchao commented:
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/user-guide/garbage-collector.md, line 15 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAA22uZA9cIwoVkF6e:-KRB4F-cADi18MJUu0mx:b9rxdp2) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L15)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> Passive voice.
> 
> "When you delete a replica set, it is often desirable for the server [editorial note: should this be "Master"?] to automatically garbage collect all the pods that the replica set has created. When you delete an owner object, the Garbage Collector (GC) deletes that owner's dependent objects.
</details>
The first sentence is one specific use case. Then I say the garbage collector fulfills this use case. And the last sentence abstracts  the function of the garbage collector. I don't think the edits express this, so I changed the paragraph differently from your suggestion. PTAL.

---

*[docs/user-guide/garbage-collector.md, line 21 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRABJjY2kGhHlHw251q:-KRB6RGueLAVWtZmSj4w:b66papo) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L21)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> "When you create a Replication Controller or ReplicaSet in Kubernetes 1.4, the Kubernetes control plane automatically sets the `ownerReference` field for each pod that set creates.
</details>
I modified it different. PTAL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
@devin-donnelly thank you! I modified some sentences differently. PTAL.

devin-donnelly commented:
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/user-guide/garbage-collector.md, line 15 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAA22uZA9cIwoVkF6e:-KRFtDlhhQCI8IVcdpi7:bnk6i0x) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L15)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary>
> The first sentence is one specific use case. Then I say the garbage collector fulfills this use case. And the last sentence abstracts  the function of the garbage collector. I don't think the edits express this, so I changed the paragraph differently from your suggestion. PTAL.
</details>
Only change I'd make is to say "When you delete, for example, a ReplicaSet"

(I'd use the actual API name "ReplicaSet" as well)

---

*[docs/user-guide/garbage-collector.md, line 17 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAAoMmdNrH1TusZilR:-KRFtVBEYWOAkvBjDFQX:bgkdry9) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L17)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> "How to establish an owner-dependent relationship between objects"
</details>
object -> objects

---

*[docs/user-guide/garbage-collector.md, line 21 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRABJjY2kGhHlHw251q:-KRFtcAlTeoX17dAVeo4:b-m51f4y) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L21)):*
<details><summary><i>Previously, caesarxuchao (Chao Xu) wrote…</i></summary>
> I modified it different. PTAL.
</details>
Still has incorrect grammar. Maybe change to "automatically sets the `ownerReference` field for each pod that the ReplicationController or ReplicaSet creates."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152)*
<!-- Sent from Reviewable.io -->

devin-donnelly commented:
Almost there. :) Thanks.

devin-donnelly commented:
*[docs/user-guide/garbage-collector.md, line 22 \[r2\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRFzL66jRAGmfzuMZka:-KRFzL66jRAGmfzuMZkb:bymlnaf) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/527cea8aa3fc4cd95a0351448a31dec21fbfd0b3/docs/user-guide/garbage-collector.md#L22)):*
> ```Markdown
> Kubernetes 1.3 added a metadata.ownerReferences field to every Kubernetes API object. If an API object is a dependent of another object, ownerReference should point to the owning API object.
> 
> When you create a replication controller or a replica set in Kubernetes 1.4, the Kubernetes control plane automatically sets the ownerReference field for each created pod to the replication controller or the replica set.
> ```

"the Kubernetes control plane automatically sets the `ownerReference` field in each created pod to point to the owning ReplicationController or ReplicaSet."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
Review status: 0 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/user-guide/garbage-collector.md, line 15 \[r1\]](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152#-KRAA22uZA9cIwoVkF6e:-KRFxpdsoTrOo8ikachx:b-dtv2i8) ([raw file](https://github.com/kubernetes/kubernetes.github.io/blob/80941d3763d769aba59459a8c4833994a3dd99aa/docs/user-guide/garbage-collector.md#L15)):*
<details><summary><i>Previously, devin-donnelly wrote…</i></summary>
> Only change I'd make is to say "When you delete, for example, a ReplicaSet"
> 
> (I'd use the actual API name "ReplicaSet" as well)
</details>
Thanks. I changed to use "ReplicaSet", "ReplicationController", "Pod" throughout the doc. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/kubernetes/kubernetes.github.io/1152)*
<!-- Sent from Reviewable.io -->

caesarxuchao commented:
Comments addressed. Thank you @devin-donnelly!

I also changed the file name from `garbage-collector.md` to `garbage-collection.md`

devin-donnelly commented:
LGTM. Would you mind fixing the merge conflict? It's preventing me from merging.

caesarxuchao commented:
@devin-donnelly sorry for the delay I was solving a P0 bug in the day. Rebased. PTAL. Thanks.

devin-donnelly commented:
Awesome, no problem. Merging in now.

lavalamp commented on docs/user-guide/garbage-collection.md:
```diff
@@ -0,0 +1,36 @@
+
+## Garbage Collection
+
+Note: the Garbage Collection is a beta feature and is enabled by default in Kubernetes version 1.4.
```
@caesarxuchao this doc may be a good place to list behavior differences with/without the garbage collector, too.

lavalamp replied:
(in an followup)
